scalding-core/src/main/scala/com/twitter/scalding/ExecutionOptimizationRules.scala

```diff
@@ -167,250 +184,69 @@ object ExecutionOptimizationRules {
-                  Some(fix(swap.map(Swap())))
-                case Some(m@Many(_, _, _)) =>
-                  def go[X, Y](m: Many[WriteExecution, X, Y, T]): FlattenedZip[WriteExecution, (A, T)] = {
-                    // we know that (X, Y) <: T
-                    // first put the write at the head, and push the old head into the second
-                    // position
-                    val many2: Many[WriteExecution, X, (A, Y), (X, (A, Y))] =
-                      Many(m.head, Many(notWrite, m.rest))
-                    // use TwistSwap to fix the types back up
-                    val twisted: FlattenedZip[WriteExecution, (A, (X, Y))] =
-                      MapZip(many2, TwistSwap[X, A, Y]())
-                    // substitute to get back to the (A, T) type finally
-                    type F[+Z] = FlattenedZip[WriteExecution, (A, Z)]
-                    m.sub.liftCo[F](twisted)
-                  }
-                  Some(fix(go(m)))
-                case Some(m@MapZip(_, _)) =>
-                  // here we just recurse without the mapFn, since removing the mapFn
-                  // reduces the size of the dag, we are making progress on the recursion
-                  def go[X](m: MapZip[WriteExecution, X, T]): Option[FlattenedZip[WriteExecution, (A, T)]] = {
-                    Many(notWrite, m.flat)
-                      .optimize
-                      .map(_.map(ZipMap.MapRight(m.mapFn)))
-                  }
-
-                  go(m).map(fix(_))
-              }
-          }
+      def map[A, B](ex: FlattenedZip[A], fn: A => B): FlattenedZip[B] = ex match {
+        case NonWrite(nonWrite) => NonWrite(nonWrite.map(fn))
+        case Write(write) => Write(mapWrite(write, fn))
+        case Composed(write, nonWrite, compose) => Composed(write, nonWrite, compose.andThen(fn))
       }
 
-      object Many {
-        /**
-         * This is a handy constructor function that builds the SubTypes instance for us
-         */
-        def apply[Ex[x] <: Execution[x], A, B](head: Ex[A], rest: FlattenedZip[Execution, B]): Many[Ex, A, B, (A, B)] =
-          Many(head, rest, SubTypes.fromSubType[(A, B), (A, B)])
+      def zip[A, B](left: FlattenedZip[A], right: FlattenedZip[B]): FlattenedZip[(A, B)] = (left, right) match {
+        case (left@NonWrite(_), right@NonWrite(_)) =>
+          NonWrite(left.nonWrite.zip(right.nonWrite))
+        case (left@NonWrite(_), right@Write(_)) =>
+          Composed(right.write, left.nonWrite, Swap[B, A]())
+        case (left@NonWrite(_), right@Composed(_, _, _)) =>
+          zipNonWriteComposed(left, right)
+
+        case (left@Write(_), right@NonWrite(_)) =>
+          Composed(left.write, right.nonWrite, Identity[(A, B)]())
+        case (left@Write(_), right@Write(_)) =>
+          Write(mergeWrite(left.write, right.write))
+        case (left@Write(_), right@Composed(_, _, _)) =>
+          zipWriteComposed(left, right)
```

Comment:
Can we inline these? I think it is harder to read to have to jump over to read the methods but they are one liners.

Reply:
Inlined everything apart from `zipWriteComposed` & `zipNonWriteComposed`. Two reasons: it used twice and inlined version doesn't go through scala's type system